### DISCUSSION TITLE
latest/how_to/save_detections/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1242] [Docs]
# latest/how_to/save_detections/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/latest/how_to/save_detections/

**Comment by jwmetrifork** [comment]
Will other export formats be implemented? YOLO, COCO, etc

**Reply by SkalskiP** [reply]
If you want to save detections in YOLO or COCO format you can do it already using [`as_coco`](https://supervision.roboflow.com/develop/datasets/core/#supervision.dataset.core.DetectionDataset.as_coco) and [`as_yolo`](https://supervision.roboflow.com/develop/datasets/core/#supervision.dataset.core.DetectionDataset.as_yolo).

**Comment by aberg2323** [comment]
Currently, I am using code like this (https://supervision.roboflow.com/latest/how_to/track_objects/#annotate-video-with-traces) to track objects. Is there any way that I can add the save the detections feature in this code so that it tracks the objects in the video saving it in the target path and saves the same detections in a csv so that the tracker_ids match in both the video and csv file?

**Reply by LinasKo** [reply]
Hi @aberg2323 :wave:

Next, instead of `sv.process_video`, you can use these tools:

1. To make it easier to get frames one-by-one, you can use the [frames generator](https://supervision.roboflow.com/utils/video/#supervision.utils.video.get_video_frames_generator).

2. To write the video, you can use [VideoSink](https://supervision.roboflow.com/utils/video/#videosink)

3. To store detections, you can use a [CSVSink](https://supervision.roboflow.com/0.19.0/detection/tools/save_detections/).

@onuralpszr has been working on some notebooks to show example usage. We still haven't added them, but you can find a nearly-done version here: https://github.com/roboflow/supervision/pull/975 - that might help with your use case.

**Comment by likith1908** [comment]
```
import supervision as sv
from ultralytics import YOLO

model = YOLO("token.pt")
frames_generator = sv.get_video_frames_generator("/home/harvestedlabs/Desktop/Codes/test.mp4")

def get_centroid(x_min, y_min, x_max, y_max):
    if x_min is None or y_min is None or x_max is None or y_max is None:
        return None, None
    x_centroid = (x_min + x_max) / 2
    y_centroid = (y_min + y_max) / 2
    return x_centroid, y_centroid

with sv.CSVSink("/home/harvestedlabs/Desktop/Codes/Likith/detections_centroid.csv") as sink:
    for frame_index, frame in enumerate(frames_generator):
        results = model(frame)[0]
        detections = sv.Detections.from_ultralytics(results)
        
        for detection in detections:
            x_min, y_min, x_max, y_max = detection[:4]
            class_name = detection[-1]
            x_centroid, y_centroid = get_centroid(x_min, y_min, x_max, y_max)
            
            if x_centroid is not None and y_centroid is not None:
                custom_data = {
                    "x_centroid": x_centroid,
                    "y_centroid": y_centroid,
                    "class_name": class_name
                }
                sink.append(detections, custom_data)

```
I am using this code to get the centre point/ centroid, but doesn't save anything in the csv file? Why is this happening ?

**Reply by LinasKo** [reply]
Hi @likith1908 👋 

I'd start debugging by printing `detections` and `custom_data` before they're appended to the sink, and then moving upwards to see what values you are getting for other variables, e.g. the centroids.

Lastly, `class_name = detection[-1]` looks suspicious - I'm not sure if it gives the class name, though I don't have the time to test myself.

**Reply by likith1908** [reply]
Thanks @LinasKo 

I'll try that.
